---
layout: splash
permalink: /
hidden: true
header:
  overlay_color: "#5e616c"
  #overlay_image: /assets/images/mm-home-page-feature.jpg
  #actions:
  #  - label: "<i class='fas fa-download'></i> Install now"
  #    url: "/docs/quick-start-guide/"
excerpt: A collection of alevin-fry quantified single-cell and single-nucleus RNA-seq datsets 
---

## Introduction

The raw data for many single-cell and single-nucleus RNA-seq experiments is publicly-available.  However, certain datasets are used _again and again_, to demonstrate data processing in tutorials, as benchmark datasets for novel methods (e.g. for clustering, dimensionality reduction, cell type identification, etc.).  In particular, 10x Genomics hosts various publicly-available datasets generated using their technology and processed via their Cell Ranger software [on their website for download](https://www.10xgenomics.com/resources/datasets).

We have created a [Nextflow](https://www.nextflow.io)-based `alevin-fry` workflow that one can use to easily quantify single-cell RNA-sequencing data in a single workflow.  The pipeline can be found [here](https://github.com/COMBINE-lab/quantaf).  To test out this initial pipeline, we have begun to reprocess the publicly-available datasets collected from the 10x website. We have focused the initial effort on standard single-cell and single-nucleus 3' gene-expression data generated using the Chromium v2 and v3 chemistries, but hope to expand the pipeline to more complex protocols soon (e.g. feature barcoding experiments) and process those data as well.  We note that **these more complex protocols can already be processed with `alevin-fry` (see the [alevin-fry tutorials](https://combine-lab.github.io/alevin-fry-tutorials/))**, but these have just not yet been incorprated into the automated Nextflow-based workflow linked above.

Additionally, to make interacting with these data as simple as possible in standard R and Python environments, we have added functions to the [`roe`](https://github.com/combine-lab/roe) and [`pyroe`](https://github.com/combine-lab/pyroe) packages to easily and programatically download these data and make them available as a [SingleCellExperiment](https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html) or [AnnData object](https://scanpy.readthedocs.io/en/latest/usage-principles.html#anndata) respectively.  See details about the [R interface](https://combine-lab.github.io/quantaf/#r-interface) and [Python interface](https://combine-lab.github.io/quantaf/#python-interface) below.

### Processed datasets:

In this section, we list the datasets we have re-processed from the [10x website](https://www.10xgenomics.com/resources/datasets) to date, as well as the link to the quantification result generated by `alevin-fry`. Note that the quantification results are stored as `tar` files that can be obtained from the corresponding Box link.

| Dataset ID   |  Dataset name |  link to re-processed data |
|--------------|:-------------|:--------------------------:|
| 1 | [500 Human PBMCs, 3' LT v3.1, Chromium Controller](https://www.10xgenomics.com/resources/datasets/500-human-pbm-cs-3-lt-v-3-1-chromium-controller-3-1-low-6-1-0) |[Box link](https://umd.box.com/shared/static/iquqxccirrsfahn3o8xpk8xoh6qdp2rf.tar)|
| 2 | [500 Human PBMCs, 3' LT v3.1, Chromium X](https://www.10xgenomics.com/resources/datasets/500-human-pbm-cs-3-lt-v-3-1-chromium-x-3-1-low-6-1-0) |[Box link](https://umd.box.com/shared/static/e54g201esm7cw0w14j4tyh6v9y2vxhyv.tar)|
| 3 | [1k PBMCs from a Healthy Donor (v3 chemistry)](https://www.10xgenomics.com/resources/datasets/1-k-pbm-cs-from-a-healthy-donor-v-3-chemistry-3-standard-3-0-0) |[Box link](https://umd.box.com/shared/static/15mgripi63srv2v442v2u3xyajq921tt.tar)|
| 4 | [10k PBMCs from a Healthy Donor (v3 chemistry)](https://www.10xgenomics.com/resources/datasets/10-k-pbm-cs-from-a-healthy-donor-v-3-chemistry-3-standard-3-0-0) |[Box link](https://umd.box.com/shared/static/01j9574g1yd93noz2pqlqjfrdhx0m1ff.tar)|
| 5 | [10k Human PBMCs, 3' v3.1, Chromium X](https://www.10xgenomics.com/resources/datasets/10k-human-pbmcs-3-ht-v3-1-chromium-x-3-1-high) |[Box link](https://umd.box.com/shared/static/pxo0nz43iiun84j5a3yslzypzd9mejan.tar)|
| 6 | [10k Human PBMCs, 3' v3.1, Chromium Controller](https://www.10xgenomics.com/resources/datasets/10k-human-pbmcs-3-v3-1-chromium-controller-3-1-high) |[Box link](https://umd.box.com/shared/static/u08hgqiailyb9grxe1ctlaa3xt6u2h4e.tar)|
| 7 | [10k Peripheral blood mononuclear cells (PBMCs) from a healthy donor, Single Indexed](https://www.10xgenomics.com/resources/datasets/10-k-peripheral-blood-mononuclear-cells-pbm-cs-from-a-healthy-donor-single-indexed-3-1-standard-4-0-0) |[Box link](https://umd.box.com/shared/static/iol9bxiv740xq6m29p2fzcoe8volsi7i.tar)|
| 8 | [10k Peripheral blood mononuclear cells (PBMCs) from a healthy donor, Dual Indexed](https://www.10xgenomics.com/resources/datasets/10-k-peripheral-blood-mononuclear-cells-pbm-cs-from-a-healthy-donor-dual-indexed-3-1-standard-4-0-0) |[Box link](https://umd.box.com/shared/static/5dzu2tw8nz9tijt8lgmelll6sbaaomh4.tar)|
| 9 | [20k Human PBMCs, 3' HT v3.1, Chromium X](https://www.10xgenomics.com/resources/datasets/20-k-human-pbm-cs-3-ht-v-3-1-chromium-x-3-1-high-6-1-0) |[Box link](https://umd.box.com/shared/static/c609sk8w6cbn4w0tcwofz4qcyjp67506.tar)|
| 10 | [PBMCs from EDTA-Treated Blood Collection Tubes Isolated via SepMate-Ficoll Gradient (3' v3.1 Chemistry)](https://www.10xgenomics.com/resources/datasets/pbmcs-3p_edta_sepmate-3-1-standard) |[Box link](https://umd.box.com/shared/static/72se7fs5t3j2cei6zth91u78kb7fg6i3.tar)|
| 11 | [PBMCs from Heparin-Treated Blood Collection Tubes Isolated via SepMate-Ficoll Gradient (3' v3.1 Chemistry)](https://www.10xgenomics.com/resources/datasets/pbmcs-3p_heparin_sepmate-3-1-standard) |[Box link](https://umd.box.com/shared/static/qc4s1qc4a6b7dsqrhck3xwla8kzgv4as.tar)|
| 12 | [PBMCs from ACD-A Treated Blood Collection Tubes Isolated via SepMate-Ficoll Gradient (3' v3.1 Chemistry)](https://www.10xgenomics.com/resources/datasets/pbmcs-3p_acda_sepmate-3-1-standard) |[Box link](https://umd.box.com/shared/static/w4207pzdj3l5iakd7d6so0fn5rt2g6mb.tar)|
| 13 | [PBMCs from Citrate-Treated Blood Collection Tubes Isolated via SepMate-Ficoll Gradient (3' v3.1 Chemistry)](https://www.10xgenomics.com/resources/datasets/pbmcs-3p_citrate_sepmate-3-1-standard) |[Box link](https://umd.box.com/shared/static/7cmuk6g3aeedto2tlbabysbj2f7hi675.tar)|
| 14 | [PBMCs from Citrate-Treated Cell Preparation Tubes (3' v3.1 Chemistry)](https://www.10xgenomics.com/resources/datasets/pbmcs-3p_citrate_cpt-3-1-standard) |[Box link](https://umd.box.com/shared/static/syzi2km1tzcw210dnyajeh1zap815lpv.tar)|
| 15 | [PBMCs from a Healthy Donor: Whole Transcriptome Analysis](https://www.10xgenomics.com/resources/datasets/pbm-cs-from-a-healthy-donor-whole-transcriptome-analysis-3-1-standard-4-0-0) |[Box link](https://umd.box.com/shared/static/dk0hmj5mpqjq56afkr5jibavy9e3yil8.tar)|
| 16 | [Whole Blood RBC Lysis for PBMCs and Neutrophils, Granulocytes, 3'](https://www.10xgenomics.com/resources/datasets/whole-blood-rbc-lysis-for-pbmcs-neutrophils-granulocytes-3-3-1-standard) |[Box link](https://umd.box.com/shared/static/9leqnwpoqn8l6kab55br3czqo72xvgg5.tar)|
| 17 | [Peripheral blood mononuclear cells (PBMCs) from a healthy donor - Manual (channel 5)](https://www.10xgenomics.com/resources/datasets/peripheral-blood-mononuclear-cells-pbm-cs-from-a-healthy-donor-manual-channel-5-3-1-standard-3-1-0) |[Box link](https://umd.box.com/shared/static/kj9i7s25cr4wdockqpro6fown2ruwjn7.tar)|
| 18 | [Peripheral blood mononuclear cells (PBMCs) from a healthy donor - Manual (channel 1)](https://www.10xgenomics.com/resources/datasets/peripheral-blood-mononuclear-cells-pbm-cs-from-a-healthy-donor-manual-channel-1-3-1-standard-3-1-0) |[Box link](https://umd.box.com/shared/static/zq4fidn62xb4uhuml3rhv2ahzmji9ujx.tar)|
| 19 | [Peripheral blood mononuclear cells (PBMCs) from a healthy donor - Chromium Connect (channel 5)](https://www.10xgenomics.com/resources/datasets/peripheral-blood-mononuclear-cells-pbm-cs-from-a-healthy-donor-chromium-connect-channel-5-3-1-standard-3-1-0) |[Box link](https://umd.box.com/shared/static/lhxagusqxru2pk2kpz4e30l1aze1chue.tar)|
| 20 | [Peripheral blood mononuclear cells (PBMCs) from a healthy donor - Chromium Connect (channel 1)](https://www.10xgenomics.com/resources/datasets/peripheral-blood-mononuclear-cells-pbm-cs-from-a-healthy-donor-chromium-connect-channel-1-3-1-standard-3-1-0) |[Box link](https://umd.box.com/shared/static/rksn46prkvukenue1ky4s77vy9sxhocy.tar)|
| 21 | [Hodgkin's Lymphoma, Dissociated Tumor: Whole Transcriptome Analysis](https://www.10xgenomics.com/resources/datasets/hodgkins-lymphoma-dissociated-tumor-whole-transcriptome-analysis-3-1-standard-4-0-0) |[Box link](https://umd.box.com/shared/static/1oj814ltwelp1ibbeq2pfbs9gyld739j.tar)|
| 22 | [200 Sorted Cells from Human Glioblastoma Multiforme, 3’ LT v3.1](https://www.10xgenomics.com/resources/datasets/200-sorted-cells-from-human-glioblastoma-multiforme-3-lt-v-3-1-3-1-low-6-0-0) |[Box link](https://umd.box.com/shared/static/ybevwqzqsjtzvg0dpaqfhwl2p386xr0r.tar)|
| 23 | [750 Sorted Cells from Human Invasive Ductal Carcinoma, 3’ LT v3.1](https://www.10xgenomics.com/resources/datasets/750-sorted-cells-from-human-invasive-ductal-carcinoma-3-lt-v-3-1-3-1-low-6-0-0) |[Box link](https://umd.box.com/shared/static/b2t6mh5mfzyr0udkmqxc55hkjqbnzora.tar)|
| 24 | [2k Sorted Cells from Human Glioblastoma Multiforme, 3’ v3.1](https://www.10xgenomics.com/resources/datasets/2-k-sorted-cells-from-human-glioblastoma-multiforme-3-v-3-1-3-1-standard-6-0-0) |[Box link](https://umd.box.com/shared/static/6y75wxp84kbs383f2z1qr8gaoxwz14ri.tar)|
| 25 | [7.5k Sorted Cells from Human Invasive Ductal Carcinoma, 3’ v3.1](https://www.10xgenomics.com/resources/datasets/7-5-k-sorted-cells-from-human-invasive-ductal-carcinoma-3-v-3-1-3-1-standard-6-0-0) |[Box link](https://umd.box.com/shared/static/ysxprpz8uhx4rdz7434792jctrx53hpp.tar)|
| 26 | [Human Glioblastoma Multiforme: 3’v3 Whole Transcriptome Analysis](https://www.10xgenomics.com/resources/datasets/human-glioblastoma-multiforme-3-v-3-whole-transcriptome-analysis-3-standard-4-0-0) |[Box link](https://umd.box.com/shared/static/n8u4wkkzc5sl1lnjik5mrsuwqdl675y1.tar)|
| 27 | [1k Brain Cells from an E18 Mouse (v3 chemistry)](https://www.10xgenomics.com/resources/datasets/1-k-brain-cells-from-an-e-18-mouse-v-3-chemistry-3-standard-3-0-0) |[Box link](https://umd.box.com/shared/static/yhvpzyc7hdnuzpd944bw7t53pyv5831m.tar)|
| 28 | [10k Brain Cells from an E18 Mouse (v3 chemistry)](https://www.10xgenomics.com/resources/datasets/10-k-brain-cells-from-an-e-18-mouse-v-3-chemistry-3-standard-3-0-0) |[Box link](https://umd.box.com/shared/static/t9ut1xc25x1ok3a7p41f529fa49gk43p.tar)|
| 29 | [1k Heart Cells from an E18 mouse (v3 chemistry)](https://www.10xgenomics.com/resources/datasets/1-k-heart-cells-from-an-e-18-mouse-v-3-chemistry-3-standard-3-0-0) |[Box link](https://umd.box.com/shared/static/9ei3jw6z1bcei7xuobuor2i5wcd1gpts.tar)|
| 30 | [10k Heart Cells from an E18 mouse (v3 chemistry)](https://www.10xgenomics.com/resources/datasets/10-k-heart-cells-from-an-e-18-mouse-v-3-chemistry-3-standard-3-0-0) |[Box link](https://umd.box.com/shared/static/hcmo9udtgp8j877ftdeapa7pzt7r60kg.tar)|
| 31 | [10k Mouse E18 Combined Cortex, Hippocampus and Subventricular Zone Cells, Single Indexed](https://www.10xgenomics.com/resources/datasets/10-k-mouse-e-18-combined-cortex-hippocampus-and-subventricular-zone-cells-single-indexed-3-1-standard-4-0-0) |[Box link](https://umd.box.com/shared/static/mr0yolo83rjdcdqgu6om4q133fpime8r.tar)|
| 32 | [10k Mouse E18 Combined Cortex, Hippocampus and Subventricular Zone Cells, Dual Indexed](https://www.10xgenomics.com/resources/datasets/10-k-mouse-e-18-combined-cortex-hippocampus-and-subventricular-zone-cells-dual-indexed-3-1-standard-4-0-0) |[Box link](https://umd.box.com/shared/static/mr7raea3v5ccn4dchemwhcimpz7t1cwl.tar)|
| 33 | [1k PBMCs from a Healthy Donor (v2 chemistry)](https://www.10xgenomics.com/resources/datasets/1-k-pbm-cs-from-a-healthy-donor-v-2-chemistry-3-standard-3-0-0) |[Box link](https://umd.box.com/shared/static/g96v4vithq75amo2yyjkesscwh6au6s4.tar)|
| 34 | [1k Brain Cells from an E18 Mouse (v2 chemistry)](https://www.10xgenomics.com/resources/datasets/1-k-brain-cells-from-an-e-18-mouse-v-2-chemistry-3-standard-3-0-0) |[Box link](https://umd.box.com/shared/static/58s10zp1lrk4pbxk9u20p3gccxp7hzi6.tar)|
| 35 | [1k Heart Cells from an E18 mouse (v2 chemistry)](https://www.10xgenomics.com/resources/datasets/1-k-heart-cells-from-an-e-18-mouse-v-2-chemistry-3-standard-3-0-0) |[Box link](https://umd.box.com/shared/static/dr8hg7fv14bgflqv97h5cpkx844ma09s.tar)|

## R and Python interface

### R interface

To ease the process of downloding the quantificaiton result of these processed datasets easily and automatically from within a programatic environment, we provide an interface in the `roe` R package exposing several useful methods. To install `roe`, please follow [this](https://github.com/COMBINE-lab/roe#installation).

- `print_available_datasets()` prints out the id and name of the available datasets.
- `get_available_dataset_df()` returns the details of the available datasets as a dataframe.
- `fetch_processed_quant(dataset_ids)` takes a vector of dataset ids as the required input, and fetches the quantification result of these datasets, according to their id, to a local directory. Other optional parameters can be found [here](https://github.com/COMBINE-lab/roe/blob/main/R/fetch_processed_quant.R#L1).
- `load_processed_quant(dataset_ids)` also takes a vector of dataset ids as the required input, and loads the quantification result of these datasets into R as `SingleCellExperiment` objects after fetching them. Other optional parameters can be found at [here](https://github.com/COMBINE-lab/roe/blob/main/R/load_processed_quant.R#L1).

The return type of both `fetch_processed_quant()` and `load_processed_quant()` is a list of [`ProcessedQuant` class instances](https://github.com/COMBINE-lab/roe/blob/main/R/ProcessedQuant.R) defined in the `roe` package. This class stores the details of a processed dataset, including the 10x chemistry, reference, dataset name, the MD5sum of the fastq.tar file, and the link to the preprocessed quantification result. It also contains the path to the fetched and decompressed quantification result, and the SingleCellExperiment object of the quantification if obtained by running `load_processed_quant(dataset_ids)`. Below we show an example:

```R
library(roe)

# fetch and decompress the quantification result of dataset #1 and #3 and load them into R as SingleCellExperiment objects.
# it returns a list of ProcessedQuant class objects, one for a fetched
# dataset
pq_list = load_processed_quant(c(1,3))

# get the ProcessedQuant object for dataset #1 and #3
pq_ds1 = pq_list[["1"]]
pq_ds3 = pq_list[["3"]]

# get the name of datset #1 and #3
pq_ds1@dataset_name

# get the link to the site storing the quantification result
pq_ds1@quant_link

# get the path to the quantification result of datset #1 and #3
pq_ds1@quant_path

# get the SingleCellExperiment object
pq_ds1@sce

# Notice that in the return object of fetch_processed_quant() 
# the sce slot is empty.
```


If one would like to fine control over the paths for saving the fetched files and decompressed folders, one can refer to the definition of the [`ProcessedQuant` class](https://github.com/COMBINE-lab/roe/blob/main/R/ProcessedQuant.R) and its functions. One can fetch, decompress and load the quntification result of a single dataset by running `ProcessedQuant()` `fetch_quant()`, `decompress_quant()` and then `load_quant()` in turn, or `FDL()`, which integrates the four functions into a single function.

### python interface

We provide analogous functions and class methods described above for fetching and loading the quantification result of datasets in the [`pyroe` python package](https://github.com/COMBINE-lab/pyroe), which can be installed via `pip` using the command:

```bash
pip install pyroe
```

or via conda using the command:

```bash
conda install -c Bioconda pyroe
```

Thanks to the flexibility of python, we offers a CLI, `pyroe fetch-quant`, for fetching and then decompressing the quantification results of any number of available datasets. The only required input is the dataset id of some available datasets. The complete list of dataset name and id is included in the help message (run `pyroe fetch-quant -h`). One can provide multiple dataset ids, separated by space, to fetch the quantification result of multiple datasets at once, for example, to fetch and decompress the quantification result of dataset #1, #3 and #6, run the command:

```bash
pyroe fetch-quant 1 3 6
```

or start python, and run

```py
import pyroe

# fetch, decompress and load the quantification result of dastset #1, 3 and 6
pq_dict = pyroe.load_processed_quant([1,3,6])

# get the ProcessedQuant class object for dataset #1 and #3
pq_ds1 = pq_dict[1]
pq_ds3 = pq_dict[3]

# get the dataset name
pq_ds1.dataset_name

# get the link to the site storing the quantification result
pq_ds1.quant_link


# get the path to the quantification result
pq_ds1.quant_path

# get the AnnData
pq_ds1.anndata

```
